packages/protocol/contracts/L1/verifiers/SgxVerifier.sol

```diff
@@ -101,59 +114,77 @@ contract SgxVerifier is EssentialContract, IVerifier {
         }
 
+        uint64 id = 0; // TODO
+
+        if (!_isIdValid(id)) revert SGX_INVALID_ID();
+        if (!_isExpired(instances[id - 1].effectiveSince)) {
+            revert SGX_INSTANCE_EXPIRED();
+        }
+
         address newInstance =
             address(bytes20(LibBytesUtils.slice(evidence.proof, 0, 20)));
-
         bytes memory signature = LibBytesUtils.slice(evidence.proof, 20);
+        bytes32 hash = getSignedHash(blockId, evidence, prover, newInstance);
 
-        address oldInstance =
-            getSignedHash(evidence, prover, newInstance).recover(signature);
-
-        _replaceInstance(oldInstance, newInstance);
-    }
-
-    function isInstanceValid(address instance) public view returns (bool) {
-        return instances[instance] + INSTANCE_EXPIRY > block.timestamp;
+        if (instances[id].addr != hash.recover(signature)) {
+            revert SGX_INVALID_AUTH();
+        }
+        _removeInstance(id);
+        _addInstance(newInstance);
+        emit InstancesChanged();
     }
 
     function getSignedHash(
+        uint64 blockId,
         TaikoData.BlockEvidence memory evidence,
         address prover,
-        address newAddress
+        address newInstance
     )
         public
         pure
         returns (bytes32 signedHash)
     {
         return keccak256(
             abi.encode(
+                blockId,
                 evidence.metaHash,
                 evidence.parentHash,
                 evidence.blockHash,
                 evidence.signalRoot,
                 evidence.graffiti,
                 prover,
-                newAddress
+                newInstance
             )
         );
     }
 
-    function _replaceInstance(
-        address oldInstance,
-        address newInstance
-    )
-        private
-    {
-        if (oldInstance != address(0)) {
-            if (!isInstanceValid(oldInstance)) revert SGX_INVALID_AUTH();
-            instances[oldInstance] = 1;
+    function _isIdValid(uint256 id) private view returns (bool) {
+        return id > 0 && id <= instances.length;
+    }
+
+    function _addInstance(address instance) private {
+        if (instance == address(0)) revert SGX_INVALID_INSTANCE();
+        instances.push(Instance(instance, uint64(block.timestamp)));
+    }
+
+    function _removeInstance(uint256 id) private onlyValidId(id) {
+        // purge up to 10 instances
+        for (uint256 i; i < 10; ++i) {
+            Instance memory last = instances[instances.length - 1];
+            if (_isExpired(last.effectiveSince)) instances.pop();
+            else break;
         }
 
-        if (newInstance == address(0)) revert SGX_INVALID_INSTANCE();
-        if (instances[newInstance] != 0) revert SGX_INSTANCE_REGISTERED();
+        if (id != instances.length) {
+            // Move the last element to the emply slot
+            Instance memory last = instances[instances.length - 1];
+            instances[id - 1] = last;
+        }
+        instances.pop();
+    }
```

Comment:
I have several problems with this.
1. How do we know which instance belongs to which ID ? (Especially if we change them with removal) Always indexing the chain ? Also SGX verifiers who wants to add with signature (`addInstanceBySgx`), needs to index for the correct id ? 
2. We want to remove Alice, she is with the 8th id, and what if there are no purged instances here?
Example:

instance.length = 10
Alice id = 8

`Instance memory last = instances[instances.length - 1];`
last is Bob
So after this:
`instances[7] = Bob;`
`instances[8] = Alice;`
`instances[9] = Dave;`

In this case why do we remove Dave ?

I feel this solution is unnecessarily complex, and would push to use this (older version) instead, as was also good with Brecht:
https://github.com/taikoxyz/taiko-mono/blob/4867aedcb0f6ae29ff8383b47da16e676bd3f253/packages/protocol/contracts/L1/verifiers/SGXVerifier.sol